### PR TITLE
neocna/milestone: model for storing azure resources

### PR DESCRIPTION
**Pull request by malkhamis-panw**
The plan will be to unmarshal the ID field of PC's JSON object that describes the azure resource. We pluck fields from it, stick it to the model, then store the whole JSON blobs. This will make the resource searchable, indexable, and retrievable.

### CODE REVIEW

File: specs/azureresource.spec [hunk 1]
```diff
@@ -0,0 +1,91 @@
+# Model
+model:
+  rest_name: azureresource
+  resource_name: azureresource
```
Comment (t00f) [inline specs/azureresource.spec:4]: resource is plural: `azureresources`

File: specs/azureresource.spec [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+      {
+        "id": "/subscriptions/.../subnets/default",
+        "name": "default",
+        "routeTableId": "/subscriptions/.../Microsoft.Network/routeTables/my-table",
+        ...
+      }
+
+  - name: kind
+    description: The specific kind of the resource.
+    type: string
```
Comment (t00f) [inline specs/azureresource.spec:41]: Are you sure you want a `string` here? We could have an `enum` so we can define all the possibilities.

File: specs/azureresource.spec [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+  - name: tags
+    description: |-
+      A contextual key-value field that can be used to narrow searching of resources
+      if the
+      resourceID is not known. For instance, it could be used to store resource
+      location or
+      public IP addresses to support cross-cloud analysis.
+    type: external
+    exposed: true
+    subtype: map[string][]string
```
Comment (t00f) [inline specs/azureresource.spec:90]: This looks weird, usually tags are an array of string, no? Not sure if we have an abstract model for this...

Comment (malkhamis-panw) [reply specs/azureresource.spec:90]: this is similar to the `annotation` field in most of our gaia models, which is `map[string][]string`. It is not like the subject/object tags in microseg.... more like `context`.. feel free to suggest another name for the field.

File: specs/azureresource.spec (hunk 2 above)
Comment (malkhamis-panw) [reply specs/azureresource.spec:41]: yes that is a good call 👍 
that also goes for the provider field, which can be Compute or Network

File: specs/azureresource.spec (hunk 3 above)
Comment (t00f) [reply specs/azureresource.spec:98]: Check the `associatedTags` in the @base. It is a list of string. Not sure what your annotations are here...

```
  - name: associatedTags
    description: List of tags attached to an entity.
    type: list
    exposed: true
    subtype: string
    stored: true
    getter: true
    setter: true
    validations:
    - $tagsWithoutReservedPrefixes
  ```

Comment (malkhamis-panw) [reply specs/azureresource.spec:98]: here is an example how we can leverage this:
```
tags: {
  "location": []{"useast"},
  "dns-names": []{"hello.world", "custom.domain"}
  }
```

this should open the door to analysis such as "_find all paths between instances located in 'uswest' to 'custom.domain'_ "...

We will still be able to efficiently find them without pulling everything or contaminating the inner logic with specifics.

File: specs/azureresource.spec [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+  - name: kind
+    description: The specific kind of the resource.
+    type: enum
+    exposed: true
+    stored: true
+    allowed_choices:
+    - VirtualMachine
+    - NetworkInterface
+    - Subnet
+    - IpConfiguration
```
Comment (primalmotion) [inline specs/azureresource.spec:48]: IPConfiguration

File: specs/azureresource.spec [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    example_value: virtualMachine
+
+  - name: name
+    description: The name of this resource.
+    type: string
+    exposed: true
+    stored: true
+    example_value: vm-name
```
Comment (primalmotion) [inline specs/azureresource.spec:56]: shouldn't that be required? You may want to use @named.abs

File: specs/azureresource.spec [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+  - name: provider
+    description: The major type of the resource.
+    type: enum
+    exposed: true
+    stored: true
+    allowed_choices:
+    - MicrosoftCompute
+    - MicrosoftNetwork
+    example_value: Microsoft.Compute
```
Comment (primalmotion) [inline specs/azureresource.spec:66]: should be required, otherwise can be empty

File: specs/azureresource.spec [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+model:
+  rest_name: azureresource
+  resource_name: azureresources
+  entity_name: AzureResource
+  package: pandemona
+  group: pcn/infrastructure
+  description: Represents an Azure cloud resource such as virtualMachines and subnets.
+  private: true
+  extends:
+  - '@migratable'
```
Comment (primalmotion) [inline specs/azureresource.spec:11]: if you plan to store this, you must make it an identifiable with @idententifiable-stored.abs

**Review by primalmotion — CHANGES_REQUESTED** [review]

File: specs/azureresource.spec (hunk 5 above)
Comment (malkhamis-panw) [reply specs/azureresource.spec:56]: yes it should be required, just like all the others.. I will take care.. thanks.

Though, `named.abs` has a cap on the char count of 256.. This name is not provided by us, but by Azure, and is the last part of the resourceID.

**Pull request merged**